# Part I: Research Question

## A.  Describe the purpose of your data mining report by doing the following:

### A1.  Propose one question relevant to a real-world organizational situation that you will answer using market basket analysis.

*What product pairings have the highest lift values in the dataset?*

### A2.  Define one goal of the data analysis. Ensure your goal is reasonable within the scope of the selected scenario and is represented in the available data.

One goal of the data analysis is to determine which item paring has the highest lift value within the dataset.

# Part II: Market Basket Justification

## B.  Explain the reasons for using market basket analysis by doing the following: 

### B1.  Explain how market basket analyzes the selected data set. Include expected outcomes.

Market basket analysis analyses a data by using the Apriori Algorithm to iteratively find the highest-scoring pairs of items within a dataset (a Pandas dataframe in this case) based on a user-selected criteria of *Lift*, *Confidence*, or *Support*.  Each of these criteria answer a slightly different question about the data depending on the question the user wants to answer in their analysis.

Modules from the *mlxtend* library in Pandas are used to first determine which rows of the original dataframe meet the criteria indicated above (including any minimum values specified by the user) using the *apriori* method, and then the  object created in that step is turned into a dataframe using the *association_rules* method within the same library. 

The outcome after using these methods is a dataframe that displays the two items that make up a particular pairing, along with the values for *Lift*, *Confidence* and *Support* (among other items).  And example can be seen in the following cell.

In [27]:
lift_rules_table[1:2]

antecedents                         consequents  \
1  (Dust-Off Compressed Gas 2 pack)  (10ft iPHone Charger Cable 2 Pack)   

   antecedent support  consequent support   support  confidence      lift  \
1            0.238368            0.050527  0.023064    0.096756  1.914955   

   leverage  conviction  zhangs_metric  
1   0.01102    1.051182        0.62733

### B2.  Provide one example of transactions in the data set.

The 4th row of the original dataframe indicates that a customer purchased the following:
* Item 01: Apple Lightning to Digital AV Adapter
* Item 02: TP-Link AC1750 Smart WiFi Router
* Item 03: Apple Pencil
    
As these were the only items purchased by the client, Items 04-20 are not applicable (and listed as *NaN* in the dataframe itself, as seen below)

In [28]:
df_orig[3:4]

Item02  \
Item01                                                                    
Apple Lightning to Digital AV Adapter  TP-Link AC1750 Smart WiFi Router   

                                             Item03 Item04 Item05 Item06  \
Item01                                                                     
Apple Lightning to Digital AV Adapter  Apple Pencil    NaN    NaN    NaN   

                                      Item07 Item08 Item09 Item10 Item11  \
Item01                                                                     
Apple Lightning to Digital AV Adapter    NaN    NaN    NaN    NaN    NaN   

                                      Item12 Item13 Item14 Item15 Item16  \
Item01                                                                     
Apple Lightning to Digital AV Adapter    NaN    NaN    NaN    NaN    NaN   

                                      Item17 Item18 Item19 Item20  
Item01                                                             
Apple Lightning to Digital AV Adapter    NaN    NaN    NaN    NaN

### B3.  Summarize one assumption of market basket analysis.

One assumption of market basket analysis is that there exists a quantifiable probability that certain items are bought together by customers (Kadlaskar, 2021).

# Part III: Data Preparation and Analysis

## C.  Prepare and perform market basket analysis by doing the following:

### C1.  Transform the data set to make it suitable for market basket analysis. Include a copy of the cleaned data set.

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

sns.set() 

import numpy as np

In [2]:
# import csv

df = pd.read_csv("C:/Users/nick_/_WGU/D212/Task III dataset/Churn/teleco_market_basket.csv", index_col=0)

# "C:\Users\nick_\_WGU\D212\Task III dataset\Churn\teleco_market_basket.csv"

In [3]:
# set notebook up to display all col names

pd.set_option('display.max_columns', None) 

In [4]:
# take look at first handful of rows

df.head(5)

Item02  \
Item01                                                                    
NaN                                                                 NaN   
Logitech M510 Wireless mouse                                  HP 63 Ink   
NaN                                                                 NaN   
Apple Lightning to Digital AV Adapter  TP-Link AC1750 Smart WiFi Router   
NaN                                                                 NaN   

                                             Item03  \
Item01                                                
NaN                                             NaN   
Logitech M510 Wireless mouse              HP 65 ink   
NaN                                             NaN   
Apple Lightning to Digital AV Adapter  Apple Pencil   
NaN                                             NaN   

                                                           Item04  \
Item01                                                              
NaN                                                           NaN   
Logitech M510 Wireless mouse           nonda USB C to USB Adapter   
NaN                                                           NaN   
Apple Lightning to Digital AV Adapter                         NaN   
NaN                                                           NaN   

                                                          Item05  \
Item01                                                             
NaN                                                          NaN   
Logitech M510 Wireless mouse           10ft iPHone Charger Cable   
NaN                                                          NaN   
Apple Lightning to Digital AV Adapter                        NaN   
NaN                                                          NaN   

                                             Item06  \
Item01                                                
NaN                                             NaN   
Logitech M510 Wireless mouse           HP 902XL ink   
NaN                                             NaN   
Apple Lightning to Digital AV Adapter           NaN   
NaN                                             NaN   

                                                             Item07  \
Item01                                                                
NaN                                                             NaN   
Logitech M510 Wireless mouse           Creative Pebble 2.0 Speakers   
NaN                                                             NaN   
Apple Lightning to Digital AV Adapter                           NaN   
NaN                                                             NaN   

                                                                    Item08  \
Item01                                                                       
NaN                                                                    NaN   
Logitech M510 Wireless mouse           Cleaning Gel Universal Dust Cleaner   
NaN                                                                    NaN   
Apple Lightning to Digital AV Adapter                                  NaN   
NaN                                                                    NaN   

                                                              Item09  \
Item01                                                                 
NaN                                                              NaN   
Logitech M510 Wireless mouse           Micro Center 32GB Memory card   
NaN                                                              NaN   
Apple Lightning to Digital AV Adapter                            NaN   
NaN                                                              NaN   

                                                                        Item10  \
Item01                                                                           
NaN                                                                        NaN   
Logitech M

In [5]:
# create a copy of the original df in case we need to reference the original values later

df_orig = df

In [6]:
# create a set of col names in case we need to reference those later

original_dataset_cols_set = set(df_orig.columns.tolist()) 

##### **Identification and treatment of null values**

In [7]:
df = df.reset_index() # thank you https://pynative.com/pandas-reset-index/
df

Item01  \
0                                        NaN   
1               Logitech M510 Wireless mouse   
2                                        NaN   
3      Apple Lightning to Digital AV Adapter   
4                                        NaN   
...                                      ...   
14997         Falcon Dust Off Compressed Gas   
14998                                    NaN   
14999                            HP 63XL Ink   
15000                                    NaN   
15001                           Apple Pencil   

                                 Item02                 Item03  \
0                                   NaN                    NaN   
1                             HP 63 Ink              HP 65 ink   
2                                   NaN                    NaN   
3      TP-Link AC1750 Smart WiFi Router           Apple Pencil   
4                                   NaN                    NaN   
...                                 ...                    ...   
14997                               NaN                    NaN   
14998                               NaN                    NaN   
14999         Apple USB-C Charger cable                    NaN   
15000                               NaN                    NaN   
15001          SanDisk Ultra 128GB card  RUNMUS Gaming Headset   

                             Item04                     Item05        Item06  \
0                               NaN                        NaN           NaN   
1        nonda USB C to USB Adapter  10ft iPHone Charger Cable  HP 902XL ink   
2                               NaN                        NaN           NaN   
3                               NaN                        NaN           NaN   
4                               NaN                        NaN           NaN   
...                             ...                        ...           ...   
14997                           NaN                        NaN           NaN   
14998                           NaN                        NaN           NaN   
14999                           NaN                        NaN           NaN   
15000                           NaN                        NaN           NaN   
15001  TopMate C5 Laptop Cooler pad                        NaN           NaN   

                             Item07                               Item08  \
0                               NaN                                  NaN   
1      Creative Pebble 2.0 Speakers  Cleaning Gel Universal Dust Cleaner   
2                               NaN                                  NaN   
3                               NaN                                  NaN   
4                               NaN                                  NaN   
...                             ...                                  ...   
14997                           NaN                                  NaN   
14998                           NaN                                  NaN   
14999                           NaN                                  NaN   
15000                           NaN                                  NaN   
15001                           NaN                                  NaN   

                              Item09                                   Item10  \
0                                NaN                                      NaN   
1      Micro Center 32GB Memory card  YUNSONG 3pack 6ft Nylon Lightning Cable   
2                                NaN                                      NaN   
3                                NaN                                      NaN   
4                                NaN                                      NaN   
...                              ...                                      ...   
14997                            NaN                                      NaN   
14998                            NaN                                      NaN   
14999                            NaN                             

In [8]:
##### **Identification and treatment of null values**

# identify col(s) with null values
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15002 entries, 0 to 15001
Data columns (total 20 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Item01  7501 non-null   object
 1   Item02  5747 non-null   object
 2   Item03  4389 non-null   object
 3   Item04  3345 non-null   object
 4   Item05  2529 non-null   object
 5   Item06  1864 non-null   object
 6   Item07  1369 non-null   object
 7   Item08  981 non-null    object
 8   Item09  654 non-null    object
 9   Item10  395 non-null    object
 10  Item11  256 non-null    object
 11  Item12  154 non-null    object
 12  Item13  87 non-null     object
 13  Item14  47 non-null     object
 14  Item15  25 non-null     object
 15  Item16  8 non-null      object
 16  Item17  4 non-null      object
 17  Item18  4 non-null      object
 18  Item19  3 non-null      object
 19  Item20  1 non-null      object
dtypes: object(20)
memory usage: 2.3+ MB


In [9]:
# view # rows before filtering out nulls

df.shape

(15002, 20)

In [10]:
# filter out all values 

df = df[df["Item01"].notna()] # thanks Dr. Kamara
df

Item01  \
1                   Logitech M510 Wireless mouse   
3          Apple Lightning to Digital AV Adapter   
5      UNEN Mfi Certified 5-pack Lightning Cable   
7                            Cat8 Ethernet Cable   
9                 Dust-Off Compressed Gas 2 pack   
...                                          ...   
14993               SanDisk 32GB Ultra SDHC card   
14995      Apple Lightning to Digital AV Adapter   
14997             Falcon Dust Off Compressed Gas   
14999                                HP 63XL Ink   
15001                               Apple Pencil   

                                     Item02                        Item03  \
1                                 HP 63 Ink                     HP 65 ink   
3          TP-Link AC1750 Smart WiFi Router                  Apple Pencil   
5                                       NaN                           NaN   
7                                 HP 65 ink                           NaN   
9             Screen Mom Screen Cleaner kit    Moread HDMI to VGA Adapter   
...                                     ...                           ...   
14993                 Vsco 70 pack stickers  SanDisk 128GB microSDXC card   
14995  Nylon Braided Lightning to USB cable                  Apple Pencil   
14997                                   NaN                           NaN   
14999             Apple USB-C Charger cable                           NaN   
15001              SanDisk Ultra 128GB card         RUNMUS Gaming Headset   

                             Item04                              Item05  \
1        nonda USB C to USB Adapter           10ft iPHone Charger Cable   
3                               NaN                                 NaN   
5                               NaN                                 NaN   
7                               NaN                                 NaN   
9             HP 62XL Tri-Color ink           Apple USB-C Charger cable   
...                             ...                                 ...   
14993                           NaN                                 NaN   
14995         USB 2.0 Printer cable  ARRIS SURFboard SB8200 Cable Modem   
14997                           NaN                                 NaN   
14999                           NaN                                 NaN   
15001  TopMate C5 Laptop Cooler pad                                 NaN   

                          Item06                        Item07  \
1                   HP 902XL ink  Creative Pebble 2.0 Speakers   
3                            NaN                           NaN   
5                            NaN                           NaN   
7                            NaN                           NaN   
9                            NaN                           NaN   
...                          ...                           ...   
14993                        NaN                           NaN   
14995  Apple USB-C Charger cable                           NaN   
14997                        NaN                           NaN   
14999                        NaN                           NaN   
15001                        NaN                           NaN   

                                    Item08                         Item09  \
1      Cleaning Gel Universal Dust Cleaner  Micro Center 32GB Memory card   
3                                      NaN                            NaN   
5                                      NaN                            NaN   
7                                      NaN                            NaN   
9                                      NaN                            NaN   
...                                    ...                            ...   
14993                                  NaN                            NaN   
14995                                  NaN                            NaN   
14997                                  NaN                            NaN   
14999                       

In [11]:
# view # rows after filtering out nulls

df.shape

(7501, 20)

#### Transforming data into a format suitable for Apriori Algorithm

Import mlxtend and relevant module

In [12]:
import mlxtend
from mlxtend.preprocessing import TransactionEncoder

Convert df into list of lists

In [13]:
import time
start = time.time()

In [14]:
rows = []
for i in range(df.shape[0]):
    rows.append([str(df.values[i, j]) for j in range(df.shape[1])]) # thanks Dr. Kamara 
    if i % 500 == 0:
        print(i)

0
500
1000
1500
2000
2500
3000
3500
4000
4500
5000
5500
6000
6500
7000
7500


In [15]:
end = time.time()

lapsed = round(end-start,2)

print(f"Time lapsed: {lapsed} seconds")

Time lapsed: 188.59 seconds


In [16]:
# feed list into TransactionEncoder

encoder = TransactionEncoder()
data_array = encoder.fit(rows).transform(rows)

In [17]:
# return output to new df

df_encoded = pd.DataFrame(data_array, columns= encoder.columns_)
df_encoded

10ft iPHone Charger Cable  10ft iPHone Charger Cable 2 Pack  \
0                          True                             False   
1                         False                             False   
2                         False                             False   
3                         False                             False   
4                         False                             False   
...                         ...                               ...   
7496                      False                             False   
7497                      False                             False   
7498                      False                             False   
7499                      False                             False   
7500                      False                             False   

      3 pack Nylon Braided Lightning Cable  3A USB Type C Cable 3 pack 6FT  \
0                                    False                            True   
1                                    False                           False   
2                                    False                           False   
3                                    False                           False   
4                                    False                           False   
...                                    ...                             ...   
7496                                 False                           False   
7497                                 False                           False   
7498                                 False                           False   
7499                                 False                           False   
7500                                 False                           False   

      5pack Nylon Braided USB C cables  ARRIS SURFboard SB8200 Cable Modem  \
0                                False                               False   
1                                False                               False   
2                                False                               False   
3                                False                               False   
4                                False                               False   
...                                ...                                 ...   
7496                             False                               False   
7497                             False                                True   
7498                             False                               False   
7499                             False                               False   
7500                             False                               False   

      Anker 2-in-1 USB Card Reader  Anker 4-port USB hub  \
0                            False                 False   
1                            False                 False   
2                            False                 False   
3                            False                 False   
4                            False                 False   
...                            ...                   ...   
7496                         False                 False   
7497                         False                 False   
7498                         False                 False   
7499                         False                 False   
7500                         False                 False   

      Anker USB C to HDMI Adapter  Apple Lightning to Digital AV Adapter  \
0                           False                                  False   
1                           False                                   True   
2                           False                                  False   
3                           False                                  False   
4                           False                                  False   
...                           ...                                    ...   
7496                        False                 

In [18]:
# remove nan row

cleaned_df_encoded = df_encoded.drop(["nan"], axis=1)
cleaned_df_encoded

10ft iPHone Charger Cable  10ft iPHone Charger Cable 2 Pack  \
0                          True                             False   
1                         False                             False   
2                         False                             False   
3                         False                             False   
4                         False                             False   
...                         ...                               ...   
7496                      False                             False   
7497                      False                             False   
7498                      False                             False   
7499                      False                             False   
7500                      False                             False   

      3 pack Nylon Braided Lightning Cable  3A USB Type C Cable 3 pack 6FT  \
0                                    False                            True   
1                                    False                           False   
2                                    False                           False   
3                                    False                           False   
4                                    False                           False   
...                                    ...                             ...   
7496                                 False                           False   
7497                                 False                           False   
7498                                 False                           False   
7499                                 False                           False   
7500                                 False                           False   

      5pack Nylon Braided USB C cables  ARRIS SURFboard SB8200 Cable Modem  \
0                                False                               False   
1                                False                               False   
2                                False                               False   
3                                False                               False   
4                                False                               False   
...                                ...                                 ...   
7496                             False                               False   
7497                             False                                True   
7498                             False                               False   
7499                             False                               False   
7500                             False                               False   

      Anker 2-in-1 USB Card Reader  Anker 4-port USB hub  \
0                            False                 False   
1                            False                 False   
2                            False                 False   
3                            False                 False   
4                            False                 False   
...                            ...                   ...   
7496                         False                 False   
7497                         False                 False   
7498                         False                 False   
7499                         False                 False   
7500                         False                 False   

      Anker USB C to HDMI Adapter  Apple Lightning to Digital AV Adapter  \
0                           False                                  False   
1                           False                                   True   
2                           False                                  False   
3                           False                                  False   
4                           False                                  False   
...                           ...                                    ...   
7496                        False                 

In [19]:
cleaned_df_encoded.shape

(7501, 119)

Save output to csv file

In [20]:
cleaned_df_encoded.to_csv("C:/Users/nick_/_WGU/D212/Task III dataset/Churn/market_basket_clean.csv")
print("Successfully saved")

Successfully saved


### C2.  Execute the code used to generate association rules with the Apriori algorithm. Provide screenshots that demonstrate that the code is error free.

In [21]:
# import relevant libraries

from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

In [22]:
# create Apriori object rules

rules = apriori(cleaned_df_encoded, min_support=0.02, use_colnames=True)
rules

support                                           itemsets
0    0.050527                 (10ft iPHone Charger Cable 2 Pack)
1    0.042528                   (3A USB Type C Cable 3 pack 6FT)
2    0.029463                     (Anker 2-in-1 USB Card Reader)
3    0.068391                      (Anker USB C to HDMI Adapter)
4    0.087188            (Apple Lightning to Digital AV Adapter)
..        ...                                                ...
98   0.023730  (USB 2.0 Printer cable, Screen Mom Screen Clea...
99   0.035462  (Screen Mom Screen Cleaner kit, VIVO Dual LCD ...
100  0.020131       (USB 2.0 Printer cable, Stylus Pen for iPad)
101  0.025197  (Stylus Pen for iPad, VIVO Dual LCD Monitor De...
102  0.027596  (USB 2.0 Printer cable, VIVO Dual LCD Monitor ...

[103 rows x 2 columns]

In [23]:
# create rules table

lift_rules_table = association_rules(rules, metric = "lift", min_threshold=1)
lift_rules_table

antecedents                         consequents  \
0   (10ft iPHone Charger Cable 2 Pack)    (Dust-Off Compressed Gas 2 pack)   
1     (Dust-Off Compressed Gas 2 pack)  (10ft iPHone Charger Cable 2 Pack)   
2        (Anker USB C to HDMI Adapter)    (Dust-Off Compressed Gas 2 pack)   
3     (Dust-Off Compressed Gas 2 pack)       (Anker USB C to HDMI Adapter)   
4   (VIVO Dual LCD Monitor Desk mount)       (Anker USB C to HDMI Adapter)   
..                                 ...                                 ...   
89  (VIVO Dual LCD Monitor Desk mount)     (Screen Mom Screen Cleaner kit)   
90             (USB 2.0 Printer cable)               (Stylus Pen for iPad)   
91               (Stylus Pen for iPad)             (USB 2.0 Printer cable)   
92               (Stylus Pen for iPad)  (VIVO Dual LCD Monitor Desk mount)   
93  (VIVO Dual LCD Monitor Desk mount)               (Stylus Pen for iPad)   

    antecedent support  consequent support   support  confidence      lift  \
0             0.050527            0.238368  0.023064    0.456464  1.914955   
1             0.238368            0.050527  0.023064    0.096756  1.914955   
2             0.068391            0.238368  0.024397    0.356725  1.496530   
3             0.238368            0.068391  0.024397    0.102349  1.496530   
4             0.174110            0.068391  0.020931    0.120214  1.757755   
..                 ...                 ...       ...         ...       ...   
89            0.174110            0.129583  0.035462    0.203675  1.571779   
90            0.170911            0.095054  0.020131    0.117785  1.239135   
91            0.095054            0.170911  0.020131    0.211781  1.239135   
92            0.095054            0.174110  0.025197    0.265077  1.522468   
93            0.174110            0.095054  0.025197    0.144717  1.522468   

    leverage  conviction  zhangs_metric  
0   0.011020    1.401255       0.503221  
1   0.011020    1.051182       0.627330  
2   0.008095    1.183991       0.356144  
3   0.008095    1.037830       0.435627  
4   0.009023    1.058905       0.521973  
..       ...         ...            ...  
89  0.012900    1.093043       0.440468  
90  0.003885    1.025766       0.232768  
91  0.003885    1.051852       0.213256  
92  0.008647    1.123778       0.379218  
93  0.008647    1.058066       0.415518  

[94 rows x 10 columns]

### C3.  Provide values for the support, lift, and confidence of the association rules table.

When the objective is to find the pairings with the greatest **Lift** value

In [24]:
rules_table_sorted = lift_rules_table.sort_values(["lift", "confidence", "support"], ascending=[False, False, False])
lift_rules_table_short = lift_rules_table[["antecedents", "consequents", "lift", "confidence", "support"]]
lift_rules_table_short

antecedents                         consequents  \
0   (10ft iPHone Charger Cable 2 Pack)    (Dust-Off Compressed Gas 2 pack)   
1     (Dust-Off Compressed Gas 2 pack)  (10ft iPHone Charger Cable 2 Pack)   
2        (Anker USB C to HDMI Adapter)    (Dust-Off Compressed Gas 2 pack)   
3     (Dust-Off Compressed Gas 2 pack)       (Anker USB C to HDMI Adapter)   
4   (VIVO Dual LCD Monitor Desk mount)       (Anker USB C to HDMI Adapter)   
..                                 ...                                 ...   
89  (VIVO Dual LCD Monitor Desk mount)     (Screen Mom Screen Cleaner kit)   
90             (USB 2.0 Printer cable)               (Stylus Pen for iPad)   
91               (Stylus Pen for iPad)             (USB 2.0 Printer cable)   
92               (Stylus Pen for iPad)  (VIVO Dual LCD Monitor Desk mount)   
93  (VIVO Dual LCD Monitor Desk mount)               (Stylus Pen for iPad)   

        lift  confidence   support  
0   1.914955    0.456464  0.023064  
1   1.914955    0.096756  0.023064  
2   1.496530    0.356725  0.024397  
3   1.496530    0.102349  0.024397  
4   1.757755    0.120214  0.020931  
..       ...         ...       ...  
89  1.571779    0.203675  0.035462  
90  1.239135    0.117785  0.020131  
91  1.239135    0.211781  0.020131  
92  1.522468    0.265077  0.025197  
93  1.522468    0.144717  0.025197  

[94 rows x 5 columns]

### C4.  Explain the top three relevant rules generated by the Apriori algorithm. Include a screenshot of the top three relevant rules.

Based on the results from the Apriori table created above, the top three rules in the dataset when evaluating for highest lift are as follows:
1. Purchasers of a *10ft iPhone Charger Cable 2 pack* buying a *Dust-Off Compressed Gas 2 pack*
2. Purchasers of a *Dust-Off Compressed Gas 2 pack* buying a *10ft iPhone Charger Cable 2 pack*
3. Purchasers of a *Dust-Off Compressed Gas 2 pack* buying an *Anker USB C to HDMI Adapter*

In [25]:
lift_rules_table_short.head(3)

antecedents                         consequents  \
0  (10ft iPHone Charger Cable 2 Pack)    (Dust-Off Compressed Gas 2 pack)   
1    (Dust-Off Compressed Gas 2 pack)  (10ft iPHone Charger Cable 2 Pack)   
2       (Anker USB C to HDMI Adapter)    (Dust-Off Compressed Gas 2 pack)   

       lift  confidence   support  
0  1.914955    0.456464  0.023064  
1  1.914955    0.096756  0.023064  
2  1.496530    0.356725  0.024397

# Part IV: Data Summary and Implications

## D.  Summarize your data analysis by doing the following:

### D1.  Summarize the significance of support, lift, and confidence from the results of the analysis.

(this analysis operates under the assumption that a customer is putting the first item into their cart, then the second item into their cart.  i.e. one overall transaction, not a customer first buying the first item and then buying the second item in a later transaction.  Both interpretations seem to be viable, so I'm stating mine here in hopes that clears up any questions in my analysis)

The significance of support, lift, and confidence from the results of the analysis are as follows: 
1. **Support** 
    * Instances of a customer adding a *10ft iPhone Charger Cable 2 pack* to their cart and then a *Dust-Off Compressed Gas 2 pack* (and vice versa) occur in **2.31%** of all transactions in the dataset
    * Instances of a customer adding a *Dust-Off Compressed Gas 2 pack* to their cart and then a *10ft iPhone Charger Cable 2 pack* occur in **2.44%** of all transactions in the dataset
    
    * ***Significance: we see that the pairing with the highest lift occurred in slightly over 2.25% of all transactions***
    
    
2. **Lift** 
    * Customers adding a *10ft iPhone Charger Cable 2 pack* to their cart are **1.91x** more likely to also add a *Dust-Off Compressed Gas 2 pack* to their cart and vice-versa
    * Customers adding a *Dust-Off Compressed Gas 2 pack* to their cart are **1.5x** more likely to also add an *Anker USB C to HDMI Adapter* to their cart
    
    * ***Significance: if a customer purchases either a cable charger pack or a compressed gas pack, the likelihood of them purchasing the other item is nearly double what it would otherwise be***
    
    
3. **Confidence** 
    * **45.6%** of all transactions with a *10ft iPhone Charger Cable 2 pack* also contained a *Dust-Off Compressed Gas 2 pack*
    * **9.7%** of transactions with a *Dust-Off Compressed Gas 2 pack* also contained a *10ft iPhone Charger Cable 2 pack*
    * **10.2%** of transactions with a *Dust-Off Compressed Gas 2 pack* also contained an *Anker USB C to HDMI Adapter*
    * ***Significance: nearly half of all transactions containing a cable charger pack also contain a compressed gas pack***

### D2.  Discuss the practical significance of your findings from the analysis.

The three rules that have the highest lift values occur in just over 2.25% of the transactions in the dataset, and the product pairs in each relationship are such that a customer is 1.5 times more likely to buy the second product if the first one is in their cart.  In the case of customers buying a *10ft iPhone Charger Cable 2 pack*, there is nearly a 1 in 2 chance that they will also buy a *Dust-Off Compressed Gas 2 pack*.  

Also of note is that single versions of the cable pack and compressed gas pack were options in the dataset.  This suggests that there is a stronger relationship between the two-packs of these items than there is between a single cable and a single compressed gas unit.

### D3.  Recommend a course of action for the real-world organizational situation from part A1 based on the results from part D1.

In light of the insights gathered here, a logical next step would be to evaluate how products are being presented wherever it is that individuals are shopping (in-store or online).  Many online retailers have a *customers also bought* section on the page of a given product.  Given the results seen here, it would make sense to build out something like that so that clients can see other products that previous customers have also bought, similar to how a scarf may be recommended when the client is on the landing page for a jacket.

# Part V: Attachments

## E.  Provide a Panopto video recording that includes the presenter and a vocalized demonstration showing all code used, the code being executed, and the results of all code used in the task.

Link can be found [here](https://wgu.hosted.panopto.com/Panopto/Pages/Viewer.aspx?id=d6899a07-f4db-4c40-adef-b0d3009b733b)

## F.  Record all web sources you used to acquire data or segments of third-party code to support the application. Ensure the web sources are reliable.

_Cited in cells where sources were used_

## G.  Acknowledge sources, using in-text citations and references, for content that is quoted, paraphrased, or summarized.

_A special thank you to Dr. Kamara for taking the time to record these videos for this Task in particular--they were most helpful!!_

Kadlaskar, A. (2021, October 2). Market basket Analysis | Guide on Market Basket Analysis. Analytics Vidhya. https://www.analyticsvidhya.com/blog/2021/10/a-comprehensive-guide-on-market-basket-analysis/

Kamara, Dr. K. (n.d.-a). Data Mining II - D212 Task 3. Retrieved December 9, 2023, from https://wgu.hosted.panopto.com/Panopto/Pages/Viewer.aspx?id=db85c4f1-0da5-4bde-a1a4-b07c0019d46d

Kamara, Dr. K. (n.d.-b). Data Mining II - D212 Theory. Retrieved December 9, 2023, from https://wgu.hosted.panopto.com/Panopto/Pages/Viewer.aspx?id=9541a29b-2f14-4c5d-9d86-af030005bcf6

Kamara, Dr. K. (n.d.-c). Installing mlxtend in anaconda on Window 10. Retrieved December 9, 2023, from https://wgu.hosted.panopto.com/Panopto/Pages/Viewer.aspx?id=339038ae-a6fd-415a-8af8-aeee00e3a0d9

Randeniya, M. (2023, April 16). Performing a Market-Basket Analysis Using the Apriori Algorithm — Python. Medium. https://medium.com/@randeniyamalitha08/performing-a-market-basket-analysis-using-the-apriori-algorithm-python-8a11742253dd

Sivek, Dr. S. C. (2020, November 17). Market Basket Analysis 101: Key Concepts. Medium. https://towardsdatascience.com/market-basket-analysis-101-key-concepts-1ddc6876cd00

Understanding Market Basket Analysis in Data Mining. (n.d.). Www.turing.com. Retrieved December 9, 2023, from https://www.turing.com/kb/market-basket-analysis

WebFOCUS 8 Technical Library. (n.d.). Infocenter.informationbuilders.com. Retrieved December 9, 2023, from https://infocenter.informationbuilders.com/wf80/index.jsp?topic=%2Fpubdocs%2FRStat16%2Fsource%2Ftopic49.htm

Working definitions (Sivek, 2020)

Support: (measuring the total number of transactions made for a particular product pairing divided by total number of transactions made)

Lift: (measuring the ratio of combined transactions to individual transactions)

Confidence: (measuring the ratio of transactions containing both items to the number of transactions with just the first item in it)


Another one (Understanding Market Basket Analysis in Data Mining, n.d.) 

Support: It is the total number of transactions made for a particular product divided by the total number of transactions made.

Confidence: It is the ratio of combined transactions to individual transactions.

Lift: It is the ratio of the confidence percent to the support percent.